### DISCUSSION TITLE
Current TypeScript conversion update+future possibilities

### DISCUSSION THREAD

**Discussion by mcyph** [maxGraph/maxGraph#51] [Announces]
Hi all (including but not limited to @junsikshim @coclav @jgadsden)!

Not sure I'll be able to work on this for the next few weeks but have managed to make a reasonable amount of progress - here's a quick status current status in case anyone has any feedback/would like to continue working on it:

* I've nearly finished converting all the main stuff to TypeScript, aside from a few files (e.g. the shapes are mostly done, whereas the layout ones I've started but need more type info filled in). The changes are merged to https://github.com/maxGraph/maxGraph/tree/mcyph-graph-refactor

* The plan to introduce tree-shaking we discussed I think still remains both the most risky and potentially lucrative part of this. This has involved trying to separate all the `mxGraph` classes into different `Graph*` files (e.g. `snap/GraphSnap.ts` etc) then making all the classes refer to each other via namespaces (e.g. `this.graph.snap.(method)`), though this is likely a temporary thing until we can come up with a plugin/object-based configurable system or similar. 
 
I can vaguely foresee using the graph something like the below mock code, based on charting libraries like [Apache eCharts](https://echarts.apache.org/en/index.html):

```typescript
import { Graph } from "maxgraph/view";
import { Rubberband } from "maxgraph/view/rubberband"; 

const graph = new Graph({
  container: (element),
  styles: {...},
  plugins: [
    new Rubberband({...rubberband options}),
  ],
});
graph.addVertex(new RectangleShape({...styles});
...
```

The reason why I think providing actual references to the modules in the plugins could be a good idea rather than strings is because explicitly importing would make tree-shaking and parameter checking with TypeScript possible. 

This is only preliminary though and would definitely like to know what other alternatives are and their pros/cons before we come to a conclusion!

**Comment by junsikshim** [comment]
I finally have some time to work on maxGraph.
It seems there are many compile errors in the codebase, so I'll begin there and eliminate errors as fast as possible basically to the point where we can run it. Then we can check against the examples as we refactor the structure.

Good work!

**Comment by junsikshim** [comment]
Just pushed my changes into the branch.

All the shapes are modified so that there are no longer nullable fields in CellStateStyles.
There were too many null checks and most of them didn't make much sense. (`this.state.opacity === null ? ...`)

For colors, `NONE` is used when there is no color assigned which is more compliant with css spec.

There might be some broken places by these changes, but I'll fix them once all the compiler errors are resolved. :)

**Reply by mcyph** [reply]
Awesome, that's fantastic! Sorry for leaving things in a broken state - things have been amazingly hectic last two weeks. I hope I can make a small amount of headway tomorrow. If I do commit I'll let you know below this comment. (If I forget to post it's probably safe to assume I haven't made any changes :P )

**Comment by junsikshim** [comment]
> The plan to introduce tree-shaking we discussed I think still remains both the most risky and potentially lucrative part of this. This has involved trying to separate all the mxGraph classes into different Graph* files (e.g. snap/GraphSnap.ts etc) then making all the classes refer to each other via namespaces (e.g. this.graph.snap.(method)), though this is likely a temporary thing until we can come up with a plugin/object-based configurable system or similar.

As I looked more into your code, I finally understood what you meant. I believe splitting Graph.ts into multiple files based on their roles is the way to go, but having nested structure/namespace might complicate things even more :(

So, I'm converting Graph* files as mixins to work with Graph.ts. Typescript support for mixins is fairly limited, so I'm implementing a workaround for it.

I'll post again when some progress is made. Thanks!

**Reply by mcyph** [reply]
Yeah I was continually debating myself whether mixins or a nested approach was best. I was wondering about whether the approach used in the Mixins heading at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes might be a viable solution.

The trouble with dynamic solutions is it might be tricky to use typescript types without making client user code more verbose. I was hoping to allow for type checking to facilitate error checking+autocompletions etc for users of the library, but I don't think this is an absolute requirement, and if it makes things unnecessarily complex we could be less strict on this front too!

**Reply by mcyph** [reply]
One approach that immediately comes to mind which might allow for typing is the spread syntax, so it could be something like e.g.

```javascript
import { graphBase } from "maxgraph/view";
import { rubberbandMixin } from "maxgraph/view/rubberbandMixin"; 

const graph = {
  ...graphBase,
  ...rubberbandMixin,
};
graph.init({
  container: (element),
  styles: {...},
  rubberband: {(rubberband options)},
});
graph.addVertex(new RectangleShape({...styles});
```

It might need to do some level of dynamic programming to make sure the constructors of each one could still be called without them being rewritten (e.g. we could add a `rubberbandConstructor` method to the `rubberbandMixin` then getting the `init` of the main graph to call everything ending in `~Constructor`). I'm not sure at this stage if this is the best solution, but this could be at least be a possibility. 

Getting the client to explicitly call `graph.initRubberband({(options)})` could also be an option, and one which could add extra checking for constructor types. One trouble I can see with this approach is that if the client doesn't call the right `init` things could fail with unintuitive errors, but there probably are a few ways of working around this too

**Reply by mcyph** [reply]
@junsikshim I haven't made any progress on the namespacing/mixins front, but letting you know I've pushed changes to fix references to the various `utils.*` functions. 

I was basically always intending this nested structure to be temporary in one form or another. I was thinking that I could convert to a nested structure first without tree-shaking, and get just a "hello world" example or similar working. After doing that, it would be easier to figure out which namespaces depend on each other, which methods should be in the base graph class, and which mixin/plugin "bundles" we could create externally. As it is right now in this "intermediate" phase, I think the graph class is in too many pieces for it to be practical, but I was having a lot of trouble considering how to go about structuring things for tree-shaking with such a big class and hope to make a diagram/spec for how this could be achieved soon!

**Reply by junsikshim** [reply]
@mcyph I just pushed what I think is the solution for splitting Graph based on their roles.
I thought it'd be best if we could use Graph methods just like `this.someFunction()` as before even if they exist in some other Graph* classes.

Before, when I was experimenting in my repo, I tried the spread syntax as you suggested, and was successful to a certain extent.
But, people use mxGraph very wildly(replacing .prototype.someFunction with custom logic and then make a mxGraph instance, and so on), it had shortcomings.

So, this time, I tried using JavaScript's prototype as a mean of extending classes. Basically I took the mixin pattern provided in the typescript handbook(https://www.typescriptlang.org/docs/handbook/mixins.html#alternative-pattern) and applied it in Graph.ts like,
```js
applyMixins(Graph, [
  GraphEvents,
  GraphImage,
  GraphCells,
  GraphSelection,
  GraphConnections,
  GraphEdge,
  GraphVertex,
  GraphOverlays,
  GraphEditing,
  GraphFolding,
  GraphLabel,
  GraphValidation,
  GraphSnap,
]);
```

However, this had problem with typescript support when I tried to access other Graph* methods/properties in one of Graph* methods(ex. Property 'xxx' does not exist on type GraphEvents). So in order to solve that, I created a type that is a mix of partial properties of other classes and extended from it.

```js
type PartialGraph = Pick<
  Graph,
  'fireEvent' | 'getDefaultParent' | 'getView' | 'getCurrentRoot' | 'getModel'
>;
type PartialCells = Pick<GraphCells, 'isCellSelectable' | 'getCells'>;
type PartialEvents = Pick<GraphEvents, 'isToggleEvent'>;
type PartialClass = PartialGraph & PartialCells & PartialEvents & EventSource;

// @ts-ignore recursive reference error
class GraphSelection extends autoImplement<PartialClass>() {
```

GraphSelection requires `fireEvent` from Graph, so we create a PartialGraph type with the selected properties.
The same is with PartialCells and PartialEvents, and then create a type via intersections called PartialClass.
We could now extend from that PartialClass(using a utility function called `autoImplement`) and have the benefit of type support for the all the dependent properties/methods in other Graph* classes.

I haven't been able to convert all the files yet, so there might be some gotchas along the way, but I think it is pretty expressive and doesn't require any other nesting structure to use it.

One small issue is that the `class GraphSelection ...` part generates an TS error saying `Type GraphSelection recursively references itself as a base type.`. But as this issue(https://github.com/microsoft/TypeScript/issues/40315) suggests, I think it is more of a warning as autocomplete works perfectly.

Thanks!

**Reply by junsikshim** [reply]
Well... I thought I was almost done with the refactoring...
I've merged and altered the Graph.prototype so that the splitted Graph* files all point to the same prototype.
But, while trying to run the HelloWorld story, I realized it was impossible to share states among Graph* classes.

Moreover, in order to enable typescript support, I had to make Graph extend `autoImplement<PartialClass>()`, but that also had problems such as Graph not able to inherit from EventSource. Modifying a prototype chain to have EventSource as a parent failed babel's `_classCallCheck`, and dodging it became more like a dirty hack.

So, I gave up the 'splitting into multiple classes and merge them somehow' approach, and began to try the old-fashioned way of `Graph.prototype.someMethod = function() {}` in Graph* files. It's not fancy and also has its own set of problems, but at least it keeps the prototype chain simple.

The most recent modified version is in the 'mcyph-graph-refactor' branch, so if someone else wants to take a shot at it, please do so.
Meanwhile, I'll keep trying different approaches.

**Reply by mcyph** [reply]
No worries @junsikshim, I think even if it's run into implementation issues conceptually I really liked your approach, I hadn't even heard about `Pick` and the other utility types. 

I haven't been able to work on maxGraph for some time due to needing to do uni work at the weekend, I likely won't be able to contribute much for the next month or so and think it's best if I leave larger design decisions to you+everyone else till then. 

One thing I was wondering is, although we could be quite dynamic with allowing for many different combinations of mixins, for a user there may only be certain combinations of mixins that make sense, like a minimal graph, one which has support for swimlanes etc, then have external plugins for things like drag-and-drop, clipboards, tooltips as separate external class instances that interact via hooks, but I haven't looked into this enough to know whether this would be practical.

**Reply by junsikshim** [reply]
Currently, I've implemented a very basic plugin system so users can choose which plugins to use when instantiating a Graph.

```js
export class Graph extends EventSource {
  constructor(
    container: HTMLElement,
    model?: Model,
    plugins: GraphPluginConstructor[] = defaultPlugins,
    stylesheet: Stylesheet | null = null
  ) {
```

and `defaultPlugins` is set to,

```js
const defaultPlugins: GraphPluginConstructor[] = [
  CellEditor,
  TooltipHandler,
  SelectionCellsHandler,
  PopupMenuHandler,
  ConnectionHandler,
  GraphHandler,
  PanningHandler,
];
```

To create a plugin, it is required to implement `GraphPlugin` and provide a static field called `pluginId`.

```js
class GraphHandler implements GraphPlugin {
  static pluginId = 'GraphHandler';

  constructor(graph: Graph) {
    this.graph = graph;
```

To retrieve a plugin instance, one can use like this,

```js
const graphHandler = this.graph.getPlugin('GraphHandler') as GraphHandler;
```

I haven't tested much with this plugin codes yet, but I think it'll work. :)

**Reply by junsikshim** [reply]
After countless hours, I finally got the storybook working again and most of the stories are somewhat working.
I'm focusing on HelloWorld story as it being the most basic example.
What I'm currently doing is basically having two sets of storybook open and comparing each states(almost line by line) to see where the difference lurks.
It's very time consuming, but it's doable. I'm just hoping there are no more core problems...

Anyone is welcome to jump on one of the stories and make everything work as it was.

**Reply by coclav** [reply]
Great work so far ! i'm trying it now

**Reply by junsikshim** [reply]
I was kind of busy since my last commit, and I won't have much time for the next couple of weeks.
It'd be great if someone else can keep working on it :)

For those who are not familiar with the progress, the current branch we are working on is `mcyph-graph-refactor`.
I think it's about time we can merge that into `development`, so let me know what you think about it.

**Reply by tbouffard** [reply]
@junsikshim @mcyph thanks for all your work on this topic. ❤️ 
I should be able to have a look at it starting from next week.

**Reply by tbouffard** [reply]
Update: unfortunately, I haven't had time to work on this so far. Hope to start soon.

**Comment by junsikshim** [comment]
I just merged `mcyph-graph-refactor` branch into `development`. 
I tried to create a PR, but couldn't make it work (have no idea why, perhaps the differences were too many...?)
Anyway, let's work on `development` from now. :)

**Reply by tbouffard** [reply]
Nice!

**Reply by junsikshim** [reply]
Just FYI, I don't have much spare time these days, and maybe only in mid-December I'll have time to contribute again.
Sorry for the delay!

**Reply by tbouffard** [reply]
No problem 😃 
I have started building and playing with your work. Lot of stuffs are already working, this is a great achievement!
